File: src/net/liplum/animations/anis/AniStateM.java
package net.liplum.animations.anis;

import arc.util.Nullable;
import mindustry.gen.Building;
import mindustry.world.Block;
import net.liplum.api.ITrigger;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

public class AniStateM<TBlock extends Block, TBuild extends Building> {
    private final TBlock block;
    private final TBuild build;
    private final AniConfig<TBlock, TBuild> config;
    @NotNull
    private AniState<TBlock, TBuild> curState;
    @Nullable
    private ISwitchAniStateListener<TBlock, TBuild> switchAniStateListener;
    @Nullable
    private Runnable onUpdate;

    public AniStateM(AniConfig<TBlock, TBuild> config, TBlock block, TBuild build) {
        this.config = config;
        this.curState = Objects.requireNonNull(config.getDefaultState());
        this.block = block;
        this.build = build;
    }

    public ISwitchAniStateListener<TBlock, TBuild> getSwitchAniStateListener() {
        return switchAniStateListener;
    }

    public void setSwitchAniStateListener(ISwitchAniStateListener<TBlock, TBuild> switchAniStateListener) {
        this.switchAniStateListener = switchAniStateListener;
    }

    public Runnable getOnUpdate() {
        return onUpdate;
    }

    public AniStateM<TBlock, TBuild> onUpdate(Runnable onUpdate) {
        this.onUpdate = onUpdate;
        return this;
    }

    @NotNull
    public AniState<TBlock, TBuild> getCurState() {
        return curState;
    }

    public void setCurState(@NotNull AniState<TBlock, TBuild> curState) {
        this.curState = curState;
    }

    public void drawBuilding() {
        curState.drawBuilding(this.build);
    }

    public boolean curOverwriteBlock() {
        return curState.isOverwriteBlock();
    }

    public void update() {
        if (onUpdate != null) {
            onUpdate.run();
        }
        for (AniState<TBlock, TBuild> to : config.getAllEntrances(curState)) {
            ITrigger<TBuild> canEnter = config.getCanEnter(curState, to);
            if (canEnter != null && canEnter.canTrigger(this.build)) {
                if (switchAniStateListener != null) {
                    switchAniStateListener.onSwitch(this.block, this.build, curState, to);
                }
                setCurState(to);
                return;
            }
        }
    }
}

File: src/net/liplum/animations/anis/AniStateM.kt
package net.liplum.animations.anis

import arc.graphics.g2d.Draw
import mindustry.gen.Building
import mindustry.world.Block

open class AniStateM<TBlock : Block, TBuild : Building>(
    val config: AniConfig<TBlock, TBuild>,
    val block: TBlock,
    val build: TBuild
) {
    val transition: TransitionEffect = config.transition
    val transitionDuration: Float = config.transitionDuration
    var curState: AniState<TBlock, TBuild> = config.defaultState!!
    var lastState: AniState<TBlock, TBuild>? = null
    var switchAniStateListener: ISwitchAniStateListener<TBlock, TBuild>? = null
    var onUpdate: Runnable? = null
        private set

    open fun onUpdate(onUpdate: Runnable?): AniStateM<TBlock, TBuild> {
        this.onUpdate = onUpdate
        return this
    }

    var lastSwitchTime = 0f
    var curTime = 0f
    open fun spend(time: Float) {
        curTime += time
    }

    open fun drawBuilding() {
        val delta = curTime - lastSwitchTime
        val progress = if (delta >= transitionDuration)
            1f
        else
            delta / transitionDuration
        transition.draw(progress, {
            lastState?.drawBuilding(build)
        }, {
            curState.drawBuilding(build)
        })
        Draw.reset()
        RESET_CONTEXT()
    }

    open fun curOverwriteBlock(): Boolean {
        return curState.isOverwriteBlock
    }

    open fun update() {
        onUpdate?.run()
        for (to in config.getAllEntrances(curState)) {
            val canEnter = config.getCanEnter(curState, to)
            if (canEnter != null && canEnter.canTrigger(build)) {
                switchAniStateListener?.onSwitch(block, build, curState, to)
                lastState = curState
                curState = to
                lastSwitchTime = curTime
                return
            }
        }
    }
}